File: 023_EqualRowAndColumnPairs.py
"""
Topic: Matrix / Hashing
Equal Row and Column Pairs: count how many rows equal columns
Time Complexity: O(n^2)
Space Complexity: O(n^2) for map
"""

from collections import Counter

def equal_pairs(grid):
    n = len(grid)
    rows = Counter(tuple(row) for row in grid)
    cols = Counter(tuple(grid[r][c] for r in range(n)) for c in range(n))
    res = 0
    for k, v in rows.items():
        res += v * cols.get(k, 0)
    return res


if __name__ == '__main__':
    print(equal_pairs([[3,2,1],[1,7,6],[2,7,7]]))  # 1
    print(equal_pairs([[1,2],[2,1]]))  # 2

File: 023_EqualRowAndColumnPairs.cpp
// Topic: Matrix / Hashing
// Equal Row and Column Pairs: Count how many (row, column) pairs in a matrix are identical.
// Time Complexity: O(n^2 * log n) or O(n^2) with hashing
// Space Complexity: O(n^2)

#include <bits/stdc++.h>
using namespace std;

int equalPairs(const vector<vector<int>>& grid)
{
    int n = grid.size();
    unordered_map<string, int> rowCount;
    rowCount.reserve(n * 2);
    for (int i = 0; i < n; ++i)
    {
        string key;
        key.reserve(n * 4);
        for (int j = 0; j < n; ++j)
        {
            key += to_string(grid[i][j]);
            key.push_back(',');
        }
        ++rowCount[key];
    }
    int ans = 0;
    for (int c = 0; c < n; ++c)
    {
        string key;
        key.reserve(n * 4);
        for (int r = 0; r < n; ++r)
        {
            key += to_string(grid[r][c]);
            key.push_back(',');
        }
        auto it = rowCount.find(key);
        if (it != rowCount.end())
            ans += it->second;
    }
    return ans;
}

int main()
{
    cout << equalPairs({{3, 2, 1}, {1, 7, 6}, {2, 7, 7}}) << '\n';  // 1
    cout << equalPairs({{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}) << '\n';  // 0
    cout << equalPairs({{1, 1, 1}, {1, 1, 1}, {1, 1, 1}}) << '\n';  // 9
    return 0;
}
